File: src/java_version/arrays/FactorialRecursion.java
package java_version.arrays;

public class FactorialRecursion {

    public static void main(String[] args) {
        System.out.println("factorial = " + calculateFactorial(5));
        System.out.println("factorial = " + calculateFactorialIteratively(5));
    }

    private static int calculateFactorial(int number) {
        if (number == 0) {
            return 1;
        }

        return number * calculateFactorial(number - 1);

    }

    private static int calculateFactorialIteratively(int number) {
        if (number == 0) {
            return 1;
        }

        int factorial = 1;

        for (int i = 1; i <= number; i++) {
            factorial *= i;
        }

        return factorial;
    }
}

File: src/kotlin_version/arrays/FactorialRecursion.kt
package kotlin_version.arrays

fun main() {
    println("Factorial: ${recursiveFactorial(6)}")
    println("Factorial: ${iterativeFactorial(6)}")
}

fun recursiveFactorial(number: Int): Int = if (number == 0) {
    1
} else {
    number * recursiveFactorial(number - 1)
}

fun iterativeFactorial(number: Int): Int {
    if (number == 0) {
        return 1
    }

    var factorial = 1

    for (i in 1..number) {
        factorial *= i
    }

    return factorial
}
